src/rust/engine/fs/src/store.rs

```diff
@@ -175,6 +234,83 @@ impl Store {
     })
   }
 
+  fn load_bytes_remote_with<T: Send + 'static, F: Fn(&[u8]) -> T + Send + 'static>(
+    &self,
+    fingerprint: Fingerprint,
+    db: Database,
+    f: Arc<Mutex<F>>,
+    cas_address: &str,
+    env: Arc<grpcio::Environment>,
+  ) -> BoxFuture<Option<T>, String> {
+    let channel = grpcio::ChannelBuilder::new(env).connect(cas_address);
+    let client = bazel_protos::bytestream_grpc::ByteStreamClient::new(channel);
+    let stream = client.read(&{
+      let mut req = bazel_protos::bytestream::ReadRequest::new();
+      // TODO: Pass a size around, or resolve that we don't need to.
+      req.set_resource_name(format!("/blobs/{}/{}", fingerprint, -1));
+      req.set_read_offset(0);
+      // 0 means no limit.
+      req.set_read_limit(0);
+      req
+    });
+
+    let store_copy = self.clone();
+    let directory_db = self.inner.directory_store;
+
+    // TODO: Work out why not wrapping this in a future::done(... .wait()) errors in gRPC channel
+    // creation.
+    future::done(stream.map(|r| r.data).concat2().wait())
+      .map(|bytes: Vec<u8>| Some(bytes))
+      .or_else(|e| match e {
+        grpcio::Error::RpcFailure(grpcio::RpcStatus {
+                                    status: grpcio::RpcStatusCode::NotFound, ..
+                                  }) => Ok(None),
+        _ => Err(format!("Error making CAS read request: {:?}", e)),
+      })
+      .and_then(move |maybe_bytes: Option<Vec<u8>>| match maybe_bytes {
+        Some(bytes) => {
+          if db == directory_db {
+            match Store::validate_directory(&bytes) {
+              Err(e) => {
+                return future::err(format!("Directory proto was not canonical: {}", e))
+                  .to_boxed() as BoxFuture<_, _>
+              }
+              _ => {}
+            }
+          }
+          store_copy
+                  .store_bytes(bytes.clone(), db)
+                  .and_then(move |retrieved_fingerprint| {
+                    if retrieved_fingerprint == fingerprint {
+                      Ok(())
+                    } else {
+                      Err(
+                        format!(
+                          "Fetched content from remote CAS but it had the wrong fingerprint \
+(wanted {} got {})",
+                          fingerprint,
+                          retrieved_fingerprint))
+                    }
+                  }).map(move |()| {
+                let f = f.lock().unwrap();
+                Some(f(&bytes))
+              }).to_boxed()
+        }
+        None => future::ok(None).to_boxed() as BoxFuture<_, _>,
```

Comment:
are the `as BoxFuture` ascriptions actually necessary...?

Reply:
Sadly so. Otherwise I'm told:
```
    = note: expected type `std::boxed::Box<futures::Map<futures::AndThen<futures_cpupool::CpuFuture<hash::Fingerprint, std::string::String>, std::result::Result<(), std::string::String>, [closure@src/store.rs:280:29: 291:20 fingerprint:_]>, [closure@src/store.rs:291:26: 291:42 result:_]>>`
               found type `std::boxed::Box<futures::FutureResult<std::option::Option<_>, _>>`
note: match arm with an incompatible type
```

This is https://github.com/rust-lang/rust/issues/46061